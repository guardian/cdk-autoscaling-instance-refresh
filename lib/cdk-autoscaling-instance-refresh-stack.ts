import { MetadataKeys } from '@guardian/cdk/lib/constants';
import { GuCertificate } from '@guardian/cdk/lib/constructs/acm';
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import {
	GuDistributionBucketParameter,
	GuLoggingStreamNameParameter,
	GuStack,
} from '@guardian/cdk/lib/constructs/core';
import { GuDnsRecordSet, RecordType } from '@guardian/cdk/lib/constructs/dns';
import {
	GuHttpsEgressSecurityGroup,
	GuVpc,
	SubnetType,
} from '@guardian/cdk/lib/constructs/ec2';
import { GuInstanceRole } from '@guardian/cdk/lib/constructs/iam';
import type { App } from 'aws-cdk-lib';
import { CfnOutput, Duration, Tags } from 'aws-cdk-lib';
import {
	AutoScalingGroup,
	GroupMetrics,
	HealthCheck,
} from 'aws-cdk-lib/aws-autoscaling';
import type { MachineImageConfig } from 'aws-cdk-lib/aws-ec2';
import {
	InstanceClass,
	InstanceSize,
	InstanceType,
	LaunchTemplate,
	OperatingSystemType,
	UserData,
} from 'aws-cdk-lib/aws-ec2';
import {
	ApplicationListener,
	ApplicationLoadBalancer,
	ApplicationProtocol,
	ApplicationTargetGroup,
	ListenerAction,
	Protocol,
	SslPolicy,
} from 'aws-cdk-lib/aws-elasticloadbalancingv2';
import { Bucket } from 'aws-cdk-lib/aws-s3';

export type BuildIdentifier =
	| 'ABC' // This version has a working healthcheck
	| 'XYZ' // This version has a working healthcheck
	| '500'; // This version has a failing healthcheck

interface CdkAutoscalingInstanceRefreshStackProps extends GuStackProps {
	app: string;
	buildIdentifier: BuildIdentifier;
}

export class CdkAutoscalingInstanceRefreshStack extends GuStack {
	constructor(
		scope: App,
		id: string,
		props: CdkAutoscalingInstanceRefreshStackProps,
	) {
		super(scope, id, props);

		const { stack, stage, app, buildIdentifier } = props;

		const filename = `${app}-${buildIdentifier}.deb`;
		const domainName = `${app}.gutools.co.uk`;

		const vpcPublicSubnets = GuVpc.subnetsFromParameter(this, {
			type: SubnetType.PUBLIC,
		});

		const vpcPrivateSubnets = GuVpc.subnetsFromParameter(this, {
			type: SubnetType.PRIVATE,
		});

		const vpc = GuVpc.fromIdParameter(this, 'vpc', {
			/*
			CDK wants privateSubnetIds to be a multiple of availabilityZones.
			We're pulling the subnets from a parameter at runtime.
			We know they evaluate to 3 subnets, but at compile time CDK doesn't.
			Set the number of AZs to 1 to avoid the error:
			  `Error: Number of privateSubnetIds (1) must be a multiple of availability zones (2).`
			 */
			availabilityZones: ['ignored'],
			publicSubnetIds: vpcPublicSubnets.map(({ subnetId }) => subnetId),
			privateSubnetIds: vpcPrivateSubnets.map(({ subnetId }) => subnetId),
		});

		const artifactBucketName =
			GuDistributionBucketParameter.getInstance(this).valueAsString;
		const bucket = Bucket.fromBucketName(
			this,
			'DistributionBucket',
			artifactBucketName,
		);
		const bucketKey = [stack, stage, app, filename].join('/');

		const userData = UserData.forLinux();
		userData.addS3DownloadCommand({
			bucket,
			bucketKey,
			localFile: `/${app}/${filename}`,
		});
		userData.addCommands(`dpkg -i /${app}/${filename}`);

		const instanceRole = new GuInstanceRole(this, { app });

		const launchTemplate = new LaunchTemplate(this, 'LaunchTemplate', {
			detailedMonitoring: true,
			instanceType: InstanceType.of(InstanceClass.T4G, InstanceSize.MICRO),
			userData,
			requireImdsv2: true,
			role: instanceRole,
			securityGroup: GuHttpsEgressSecurityGroup.forVpc(this, { app, vpc }),
			instanceMetadataTags: true,
			machineImage: {
				getImage: (): MachineImageConfig => {
					return {
						osType: OperatingSystemType.LINUX,
						userData,
						imageId: 'ami-0ba2d40d73e6cf972',
					};
				},
			},
		});

		new CfnOutput(this, `LaunchTemplateId-${app}`, {
			key: `LaunchTemplateId-${app}`.replaceAll('-', ''),
			value: launchTemplate.launchTemplateId!,
		});

		new CfnOutput(this, `LaunchTemplateLatestVersion-${app}`, {
			key: `LaunchTemplateLatestVersion-${app}`.replaceAll('-', ''),
			value: launchTemplate.latestVersionNumber,
		});

		const autoscalingGroup = new AutoScalingGroup(this, 'AutoScalingGroup', {
			vpc,
			vpcSubnets: {
				subnets: vpc.privateSubnets,
			},
			minCapacity: 3,
			maxCapacity: 10,
			launchTemplate,
			groupMetrics: [GroupMetrics.all()],
			healthCheck: HealthCheck.elb({
				// How long before the ASG checks that the ALB thinks the instance is healthy
				grace: Duration.minutes(2),
			}),
			defaultInstanceWarmup: Duration.seconds(0),
		});

		// Tags for application log shipping
		Tags.of(autoscalingGroup).add(
			MetadataKeys.LOG_KINESIS_STREAM_NAME,
			GuLoggingStreamNameParameter.getInstance(this).valueAsString,
		);
		Tags.of(autoscalingGroup).add(MetadataKeys.SYSTEMD_UNIT, `${app}.service`);

		new CfnOutput(this, 'AutoScalingGroupName', {
			key: 'AutoScalingGroup',
			value: autoscalingGroup.autoScalingGroupName,
		});

		const targetGroup = new ApplicationTargetGroup(this, 'TargetGroup', {
			protocol: ApplicationProtocol.HTTP,
			deregistrationDelay: Duration.seconds(30),
			vpc,
			healthCheck: {
				path: '/healthcheck',
				protocol: Protocol.HTTP,
				healthyThresholdCount: 5,
				unhealthyThresholdCount: 2,
				interval: Duration.seconds(10),
				timeout: Duration.seconds(2),
			},
			port: 9000,
			targets: [autoscalingGroup],
		});

		const loadBalancer = new ApplicationLoadBalancer(this, 'ALB', {
			vpc,
			vpcSubnets: {
				subnets: vpc.publicSubnets,
			},
			internetFacing: true,
		});
		const certificate = new GuCertificate(this, {
			app,
			domainName,
		});

		new ApplicationListener(this, 'Listener', {
			loadBalancer,
			defaultAction: ListenerAction.forward([targetGroup]),
			port: 443,
			protocol: ApplicationProtocol.HTTPS,
			sslPolicy: SslPolicy.RECOMMENDED_TLS,
			certificates: [certificate],
		});

		new GuDnsRecordSet(this, 'DNS', {
			name: domainName,
			ttl: Duration.minutes(1),
			recordType: RecordType.CNAME,
			resourceRecords: [loadBalancer.loadBalancerDnsName],
		});

		autoscalingGroup.scaleOnRequestCount('ScaleOutOnRequests', {
			targetRequestsPerMinute: 5,
		});
	}
}
