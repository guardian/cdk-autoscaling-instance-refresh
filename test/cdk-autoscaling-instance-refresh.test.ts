import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { CdkAutoscalingInstanceRefreshStack } from '../lib/cdk-autoscaling-instance-refresh-stack';

describe('CdkAutoscalingInstanceRefreshStack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new CdkAutoscalingInstanceRefreshStack(
			app,
			'CdkAutoscalingInstanceRefreshStack',
			{
				stack: 'playground',
				stage: 'PROD',
				app: 'asg-instance-refresh',
				env: {
					region: 'eu-west-1',
				},
				buildIdentifier: 'ABC',
			},
		);
		expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
	});
});
