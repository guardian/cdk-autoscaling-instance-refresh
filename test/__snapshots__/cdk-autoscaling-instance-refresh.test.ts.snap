// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdkAutoscalingInstanceRefreshStack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuSubnetListParameter",
      "GuSubnetListParameter",
      "GuVpcParameter",
      "GuDistributionBucketParameter",
      "GuInstanceRole",
      "GuSsmSshPolicy",
      "GuDescribeEC2Policy",
      "GuLoggingStreamNameParameter",
      "GuLogShippingPolicy",
      "GuGetDistributablePolicy",
      "GuParameterStoreReadPolicy",
      "GuHttpsEgressSecurityGroup",
      "GuCertificate",
      "GuDnsRecordSet",
    ],
    "gu:cdk:version": "59.2.3",
  },
  "Outputs": {
    "AutoScalingGroup": {
      "Value": {
        "Ref": "AutoScalingGroupASG804C35BE",
      },
    },
    "LaunchTemplateIdasginstancerefresh": {
      "Value": {
        "Ref": "LaunchTemplate04EC5460",
      },
    },
    "LaunchTemplateLatestVersionasginstancerefresh": {
      "Value": {
        "Fn::GetAtt": [
          "LaunchTemplate04EC5460",
          "LatestVersionNumber",
        ],
      },
    },
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "LoggingStreamName": {
      "Default": "/account/services/logging.stream.name",
      "Description": "SSM parameter containing the Name (not ARN) on the kinesis stream",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "PrivateSubnets": {
      "Default": "/account/vpc/primary/subnets/private",
      "Description": "A list of private subnets",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
    "PublicSubnets": {
      "Default": "/account/vpc/primary/subnets/public",
      "Description": "A list of public subnets",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
    "VpcId": {
      "Default": "/account/vpc/primary/id",
      "Description": "Virtual Private Cloud to run EC2 instances within. Should NOT be the account default VPC.",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::VPC::Id>",
    },
  },
  "Resources": {
    "ALBAEE750D2": {
      "Properties": {
        "LoadBalancerAttributes": [
          {
            "Key": "deletion_protection.enabled",
            "Value": "false",
          },
        ],
        "Scheme": "internet-facing",
        "SecurityGroups": [
          {
            "Fn::GetAtt": [
              "ALBSecurityGroup8B8624F8",
              "GroupId",
            ],
          },
        ],
        "Subnets": {
          "Ref": "PublicSubnets",
        },
        "Tags": [
          {
            "Key": "App",
            "Value": "asg-instance-refresh",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "59.2.3",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk-autoscaling-instance-refresh",
          },
          {
            "Key": "Stack",
            "Value": "playground",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "ALBSecurityGroup8B8624F8": {
      "Properties": {
        "GroupDescription": "Automatically created Security Group for ELB CdkAutoscalingInstanceRefreshStackALB4107979D",
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow from anyone on port 443",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "asg-instance-refresh",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "59.2.3",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk-autoscaling-instance-refresh",
          },
          {
            "Key": "Stack",
            "Value": "playground",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "VpcId": {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "ALBSecurityGrouptoCdkAutoscalingInstanceRefreshStackGuHttpsEgressSecurityGroupAsginstancerefreshBCB885AC9000F0AB3140": {
      "Properties": {
        "Description": "Load balancer to target",
        "DestinationSecurityGroupId": {
          "Fn::GetAtt": [
            "GuHttpsEgressSecurityGroupAsginstancerefresh353108F9",
            "GroupId",
          ],
        },
        "FromPort": 9000,
        "GroupId": {
          "Fn::GetAtt": [
            "ALBSecurityGroup8B8624F8",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "ToPort": 9000,
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
    "AutoScalingGroupASG804C35BE": {
      "Properties": {
        "DefaultInstanceWarmup": 0,
        "HealthCheckGracePeriod": 120,
        "HealthCheckType": "ELB",
        "LaunchTemplate": {
          "LaunchTemplateId": {
            "Ref": "LaunchTemplate04EC5460",
          },
          "Version": {
            "Fn::GetAtt": [
              "LaunchTemplate04EC5460",
              "LatestVersionNumber",
            ],
          },
        },
        "MaxSize": "10",
        "MetricsCollection": [
          {
            "Granularity": "1Minute",
          },
        ],
        "MinSize": "3",
        "Tags": [
          {
            "Key": "App",
            "PropagateAtLaunch": true,
            "Value": "asg-instance-refresh",
          },
          {
            "Key": "gu:cdk:version",
            "PropagateAtLaunch": true,
            "Value": "59.2.3",
          },
          {
            "Key": "gu:repo",
            "PropagateAtLaunch": true,
            "Value": "guardian/cdk-autoscaling-instance-refresh",
          },
          {
            "Key": "LogKinesisStreamName",
            "PropagateAtLaunch": true,
            "Value": {
              "Ref": "LoggingStreamName",
            },
          },
          {
            "Key": "Stack",
            "PropagateAtLaunch": true,
            "Value": "playground",
          },
          {
            "Key": "Stage",
            "PropagateAtLaunch": true,
            "Value": "PROD",
          },
          {
            "Key": "SystemdUnit",
            "PropagateAtLaunch": true,
            "Value": "asg-instance-refresh.service",
          },
        ],
        "TargetGroupARNs": [
          {
            "Ref": "TargetGroup3D7CD9B8",
          },
        ],
        "TerminationPolicies": [
          "NewestInstance",
        ],
        "VPCZoneIdentifier": {
          "Ref": "PrivateSubnets",
        },
      },
      "Type": "AWS::AutoScaling::AutoScalingGroup",
      "UpdatePolicy": {
        "AutoScalingScheduledAction": {
          "IgnoreUnmodifiedGroupSizeProperties": true,
        },
      },
    },
    "AutoScalingGroupScalingPolicyScaleOutOnRequests25C2FA91": {
      "DependsOn": [
        "Listener828B0E81",
      ],
      "Properties": {
        "AutoScalingGroupName": {
          "Ref": "AutoScalingGroupASG804C35BE",
        },
        "PolicyType": "TargetTrackingScaling",
        "TargetTrackingConfiguration": {
          "PredefinedMetricSpecification": {
            "PredefinedMetricType": "ALBRequestCountPerTarget",
            "ResourceLabel": {
              "Fn::Join": [
                "",
                [
                  {
                    "Fn::Select": [
                      1,
                      {
                        "Fn::Split": [
                          "/",
                          {
                            "Ref": "Listener828B0E81",
                          },
                        ],
                      },
                    ],
                  },
                  "/",
                  {
                    "Fn::Select": [
                      2,
                      {
                        "Fn::Split": [
                          "/",
                          {
                            "Ref": "Listener828B0E81",
                          },
                        ],
                      },
                    ],
                  },
                  "/",
                  {
                    "Fn::Select": [
                      3,
                      {
                        "Fn::Split": [
                          "/",
                          {
                            "Ref": "Listener828B0E81",
                          },
                        ],
                      },
                    ],
                  },
                  "/",
                  {
                    "Fn::GetAtt": [
                      "TargetGroup3D7CD9B8",
                      "TargetGroupFullName",
                    ],
                  },
                ],
              ],
            },
          },
          "TargetValue": 5,
        },
      },
      "Type": "AWS::AutoScaling::ScalingPolicy",
    },
    "CertificateAsginstancerefresh39421E9B": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "DomainName": "asg-instance-refresh.gutools.co.uk",
        "Tags": [
          {
            "Key": "App",
            "Value": "asg-instance-refresh",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "59.2.3",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk-autoscaling-instance-refresh",
          },
          {
            "Key": "Name",
            "Value": "CdkAutoscalingInstanceRefreshStack/CertificateAsginstancerefresh",
          },
          {
            "Key": "Stack",
            "Value": "playground",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "ValidationMethod": "DNS",
      },
      "Type": "AWS::CertificateManager::Certificate",
      "UpdateReplacePolicy": "Retain",
    },
    "DNS": {
      "Properties": {
        "Name": "asg-instance-refresh.gutools.co.uk",
        "RecordType": "CNAME",
        "ResourceRecords": [
          {
            "Fn::GetAtt": [
              "ALBAEE750D2",
              "DNSName",
            ],
          },
        ],
        "Stage": "PROD",
        "TTL": 60,
      },
      "Type": "Guardian::DNS::RecordSet",
    },
    "DescribeEC2PolicyFF5F9295": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "autoscaling:DescribeAutoScalingInstances",
                "autoscaling:DescribeAutoScalingGroups",
                "ec2:DescribeTags",
                "ec2:DescribeInstances",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "describe-ec2-policy",
        "Roles": [
          {
            "Ref": "InstanceRoleAsginstancerefresh239E41A5",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GetDistributablePolicyAsginstancerefresh80B91939": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:s3:::",
                    {
                      "Ref": "DistributionBucketName",
                    },
                    "/playground/PROD/asg-instance-refresh/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GetDistributablePolicyAsginstancerefresh80B91939",
        "Roles": [
          {
            "Ref": "InstanceRoleAsginstancerefresh239E41A5",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GuHttpsEgressSecurityGroupAsginstancerefresh353108F9": {
      "Properties": {
        "GroupDescription": "Allow all outbound HTTPS traffic",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound HTTPS traffic",
            "FromPort": 443,
            "IpProtocol": "tcp",
            "ToPort": 443,
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "asg-instance-refresh",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "59.2.3",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk-autoscaling-instance-refresh",
          },
          {
            "Key": "Stack",
            "Value": "playground",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "VpcId": {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "GuHttpsEgressSecurityGroupAsginstancerefreshfromCdkAutoscalingInstanceRefreshStackALBSecurityGroup078888369000E6B2F97D": {
      "Properties": {
        "Description": "Load balancer to target",
        "FromPort": 9000,
        "GroupId": {
          "Fn::GetAtt": [
            "GuHttpsEgressSecurityGroupAsginstancerefresh353108F9",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": {
          "Fn::GetAtt": [
            "ALBSecurityGroup8B8624F8",
            "GroupId",
          ],
        },
        "ToPort": 9000,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "GuLogShippingPolicy981BFE5A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "kinesis:Describe*",
                "kinesis:Put*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:kinesis:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":stream/",
                    {
                      "Ref": "LoggingStreamName",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GuLogShippingPolicy981BFE5A",
        "Roles": [
          {
            "Ref": "InstanceRoleAsginstancerefresh239E41A5",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "InstanceRoleAsginstancerefresh239E41A5": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Path": "/",
        "Tags": [
          {
            "Key": "App",
            "Value": "asg-instance-refresh",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "59.2.3",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk-autoscaling-instance-refresh",
          },
          {
            "Key": "Stack",
            "Value": "playground",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LaunchTemplate04EC5460": {
      "DependsOn": [
        "InstanceRoleAsginstancerefresh239E41A5",
      ],
      "Properties": {
        "LaunchTemplateData": {
          "IamInstanceProfile": {
            "Arn": {
              "Fn::GetAtt": [
                "LaunchTemplateProfile94AA77CE",
                "Arn",
              ],
            },
          },
          "ImageId": "ami-0ba2d40d73e6cf972",
          "InstanceType": "t4g.micro",
          "MetadataOptions": {
            "HttpTokens": "required",
            "InstanceMetadataTags": "enabled",
          },
          "Monitoring": {
            "Enabled": true,
          },
          "SecurityGroupIds": [
            {
              "Fn::GetAtt": [
                "GuHttpsEgressSecurityGroupAsginstancerefresh353108F9",
                "GroupId",
              ],
            },
          ],
          "TagSpecifications": [
            {
              "ResourceType": "instance",
              "Tags": [
                {
                  "Key": "App",
                  "Value": "asg-instance-refresh",
                },
                {
                  "Key": "gu:cdk:version",
                  "Value": "59.2.3",
                },
                {
                  "Key": "gu:repo",
                  "Value": "guardian/cdk-autoscaling-instance-refresh",
                },
                {
                  "Key": "Name",
                  "Value": "CdkAutoscalingInstanceRefreshStack/LaunchTemplate",
                },
                {
                  "Key": "Stack",
                  "Value": "playground",
                },
                {
                  "Key": "Stage",
                  "Value": "PROD",
                },
              ],
            },
            {
              "ResourceType": "volume",
              "Tags": [
                {
                  "Key": "App",
                  "Value": "asg-instance-refresh",
                },
                {
                  "Key": "gu:cdk:version",
                  "Value": "59.2.3",
                },
                {
                  "Key": "gu:repo",
                  "Value": "guardian/cdk-autoscaling-instance-refresh",
                },
                {
                  "Key": "Name",
                  "Value": "CdkAutoscalingInstanceRefreshStack/LaunchTemplate",
                },
                {
                  "Key": "Stack",
                  "Value": "playground",
                },
                {
                  "Key": "Stage",
                  "Value": "PROD",
                },
              ],
            },
          ],
          "UserData": {
            "Fn::Base64": {
              "Fn::Join": [
                "",
                [
                  "#!/bin/bash
mkdir -p $(dirname '/asg-instance-refresh/asg-instance-refresh-ABC.deb')
aws s3 cp 's3://",
                  {
                    "Ref": "DistributionBucketName",
                  },
                  "/playground/PROD/asg-instance-refresh/asg-instance-refresh-ABC.deb' '/asg-instance-refresh/asg-instance-refresh-ABC.deb'
dpkg -i /asg-instance-refresh/asg-instance-refresh-ABC.deb",
                ],
              ],
            },
          },
        },
        "TagSpecifications": [
          {
            "ResourceType": "launch-template",
            "Tags": [
              {
                "Key": "App",
                "Value": "asg-instance-refresh",
              },
              {
                "Key": "gu:cdk:version",
                "Value": "59.2.3",
              },
              {
                "Key": "gu:repo",
                "Value": "guardian/cdk-autoscaling-instance-refresh",
              },
              {
                "Key": "Name",
                "Value": "CdkAutoscalingInstanceRefreshStack/LaunchTemplate",
              },
              {
                "Key": "Stack",
                "Value": "playground",
              },
              {
                "Key": "Stage",
                "Value": "PROD",
              },
            ],
          },
        ],
      },
      "Type": "AWS::EC2::LaunchTemplate",
    },
    "LaunchTemplateProfile94AA77CE": {
      "Properties": {
        "Roles": [
          {
            "Ref": "InstanceRoleAsginstancerefresh239E41A5",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "Listener828B0E81": {
      "Properties": {
        "Certificates": [
          {
            "CertificateArn": {
              "Ref": "CertificateAsginstancerefresh39421E9B",
            },
          },
        ],
        "DefaultActions": [
          {
            "TargetGroupArn": {
              "Ref": "TargetGroup3D7CD9B8",
            },
            "Type": "forward",
          },
        ],
        "LoadBalancerArn": {
          "Ref": "ALBAEE750D2",
        },
        "Port": 443,
        "Protocol": "HTTPS",
        "SslPolicy": "ELBSecurityPolicy-TLS13-1-2-2021-06",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "ParameterStoreReadAsginstancerefreshD5CF2975": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/playground/asg-instance-refresh",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:eu-west-1:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/playground/asg-instance-refresh/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "parameter-store-read-policy",
        "Roles": [
          {
            "Ref": "InstanceRoleAsginstancerefresh239E41A5",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "SsmSshPolicy4CFC977E": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "ec2messages:AcknowledgeMessage",
                "ec2messages:DeleteMessage",
                "ec2messages:FailMessage",
                "ec2messages:GetEndpoint",
                "ec2messages:GetMessages",
                "ec2messages:SendReply",
                "ssm:UpdateInstanceInformation",
                "ssm:ListInstanceAssociations",
                "ssm:DescribeInstanceProperties",
                "ssm:DescribeDocumentParameters",
                "ssmmessages:CreateControlChannel",
                "ssmmessages:CreateDataChannel",
                "ssmmessages:OpenControlChannel",
                "ssmmessages:OpenDataChannel",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ssm-ssh-policy",
        "Roles": [
          {
            "Ref": "InstanceRoleAsginstancerefresh239E41A5",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TargetGroup3D7CD9B8": {
      "Properties": {
        "HealthCheckIntervalSeconds": 10,
        "HealthCheckPath": "/healthcheck",
        "HealthCheckProtocol": "HTTP",
        "HealthCheckTimeoutSeconds": 2,
        "HealthyThresholdCount": 5,
        "Port": 9000,
        "Protocol": "HTTP",
        "Tags": [
          {
            "Key": "App",
            "Value": "asg-instance-refresh",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "59.2.3",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk-autoscaling-instance-refresh",
          },
          {
            "Key": "Stack",
            "Value": "playground",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "TargetGroupAttributes": [
          {
            "Key": "deregistration_delay.timeout_seconds",
            "Value": "30",
          },
          {
            "Key": "stickiness.enabled",
            "Value": "false",
          },
        ],
        "TargetType": "instance",
        "UnhealthyThresholdCount": 2,
        "VpcId": {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
    },
  },
}
`;
